akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -359,127 +359,129 @@ object ShardCoordinator {
+      log.debug("The coordinator state was successfully updated with {}", evt)
       evt match {
         case ShardRegionRegistered(region) ⇒
-          persistentState = persistentState.updated(evt)
-        case ShardRegionProxyRegistered(proxy) ⇒
-          persistentState = persistentState.updated(evt)
-        case ShardRegionTerminated(region) ⇒
-          if (persistentState.regions.contains(region))
-            persistentState = persistentState.updated(evt)
-          else {
-            log.debug("ShardRegionTerminated, but region {} was not registered. This inconsistency is due to that " +
-              " some stored ActorRef in Akka v2.3.0 and v2.3.1 did not contain full address information. It will be " +
-              "removed by later watch.", region)
-          }
-        case ShardRegionProxyTerminated(proxy) ⇒
-          if (persistentState.regionProxies.contains(proxy))
-            persistentState = persistentState.updated(evt)
-        case ShardHomeAllocated(shard, region) ⇒
-          persistentState = persistentState.updated(evt)
-        case _: ShardHomeDeallocated ⇒
-          persistentState = persistentState.updated(evt)
+          val firstRegion = state.regions.isEmpty
+
+          state = state.updated(evt)
+          context.watch(region)
+          region ! RegisterAck(self)
+
+          if (firstRegion)
+            allocateShardHomes()
+        case _ ⇒
       }
 
-    case SnapshotOffer(_, state: State) ⇒
-      log.debug("receiveRecover SnapshotOffer {}", state)
-      //Old versions of the state object may not have unallocatedShard set,
-      // thus it will be null.
-      if (state.unallocatedShards == null)
-        persistentState = state.copy(unallocatedShards = Set.empty)
-      else
-        persistentState = state
-
-    case RecoveryCompleted ⇒
-      persistentState.regionProxies.foreach(context.watch)
-      persistentState.regions.foreach { case (a, _) ⇒ context.watch(a) }
-      persistentState.shards.foreach { case (a, r) ⇒ sendHostShardMsg(a, r) }
-      allocateShardHomes()
-  }
+    case UpdateTimeout(CoordinatorStateKey, Some(evt: DomainEvent)) ⇒
+      evt match {
+        case ShardRegionRegistered(region) ⇒
+          log.warning(
+            "The ShardCoordinator was unable to update a distributed state within 'updating-state-timeout'={} millis, event={}",
+            updatingStateTimeout.toMillis,
+            evt)
+        case _ ⇒
+          log.error(
+            "The ShardCoordinator was unable to update a distributed state within 'updating-state-timeout'={} millis (was retrying), event={}",
+            updatingStateTimeout.toMillis,
+            evt)
+          sendUpdate(evt)
+      }
+
+    case ModifyFailure(CoordinatorStateKey, error, cause, Some(evt: DomainEvent)) ⇒
+      log.error(
+        cause,
+        "The ShardCoordinator was unable to update a distributed state with error {}. Was retrying with event {}.",
+        error,
+        evt)
+      sendUpdate(evt)
```

Comment:
Could it happens due to message/updates reordering? And would fixed after some time.